### PR TITLE
FLW-655 Validator info screen

### PR DESCRIPTION
**Pull request by stas-litvinov**
Add empty stake screen
Add bottom sheet for Total stake
Add nomination state
Add price information
Minor design changes

### CODE REVIEW

File: Podfile [hunk 1]
```diff
@@ -16,7 +16,7 @@ abstract_target 'fearlessAll' do
   pod 'ReachabilitySwift'
   pod 'SwiftGifOrigin', '~> 1.7.0'
   pod 'SnapKit', '~> 5.0.0'
-  pod 'SwiftFormat/CLI'
+  pod 'SwiftFormat/CLI', '~> 0.47.13'
```
Comment (ef1rspb) [inline Podfile:19]: run `pod install` instead of `pod update`

File: fearless/Modules/Staking/ValidatorInfo/ValidatorInfoInteractor.swift [hunk 2]
```diff
@@ -1,17 +1,61 @@
...
+        validatorInfo: ValidatorInfoProtocol,
+        priceProvider: AnySingleValueProvider<PriceData>
+    ) {
         self.validatorInfo = validatorInfo
+        self.priceProvider = priceProvider
+    }
+
+    private func subscribeToPriceChanges() {
+        let updateClosure = { [weak self] (changes: [DataProviderChange<PriceData>]) in
+            if changes.isEmpty {
```
Comment (ef1rspb) [inline fearless/Modules/Staking/ValidatorInfo/ValidatorInfoInteractor.swift:20]: suggest `let priceData = changes.reduceToLastChange()`

File: fearless/Modules/Staking/ValidatorInfo/ValidatorInfoInteractor.swift [hunk 3]
```diff
@@ -1,17 +1,61 @@
...
+import RobinHood
 
 final class ValidatorInfoInteractor {
-    weak var presenter: ValidatorInfoInteractorOutputProtocol?
-    var validatorInfo: ValidatorInfoProtocol?
+    weak var presenter: ValidatorInfoInteractorOutputProtocol!
+    private let priceProvider: AnySingleValueProvider<PriceData>
 
-    init(validatorInfo: ValidatorInfoProtocol) {
+    private var validatorInfo: ValidatorInfoProtocol?
```
Comment (ef1rspb) [inline fearless/Modules/Staking/ValidatorInfo/ValidatorInfoInteractor.swift:8]: let?

File: fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift [hunk 4]
```diff
@@ -2,32 +2,7 @@ import UIKit
...
-    enum SectionType: Int, CaseIterable {
-        static let rowHeight: CGFloat = 52.0
-
-        case staking
-        case identity
-    }
-
-    var accountViewModel: ValidatorInfoAccountViewModelProtocol?
-    var extrasViewModel: [Section] = []
+    var viewModel: [ValidatorInfoViewModel] = []
```
Comment (ef1rspb) [inline fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift:5]: viewModels?

File: fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift [hunk 5]
```diff
@@ -43,74 +18,90 @@ final class ValidatorInfoViewController: UIViewController {
...
 
 extension ValidatorInfoViewController: UITableViewDelegate {
-    func tableView(_: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
-        guard indexPath.section > 0 else { return RowType.accountRowHeight }
+    func tableView(_: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+        switch viewModel[section] {
+        case .account:
+            return CGFloat.leastNormalMagnitude
+        default:
+            return ValidatorInfoViewModel.headerHeight
```
Comment (ef1rspb) [inline fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift:43]: viewModel knows about view height? hmm

File: fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift [hunk 6]
```diff
@@ -43,74 +18,90 @@ final class ValidatorInfoViewController: UIViewController {
...
+        }
+    }
 
-        return RowType.rowHeight
+    func tableView(_: UITableView, heightForFooterInSection _: Int) -> CGFloat {
+        CGFloat.leastNormalMagnitude
+    }
+
+    func tableView(_: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
```
Comment (ef1rspb) [inline fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift:51]: non AutoLayout cells?

File: fearless/Modules/Staking/ValidatorInfo/View/ValidatorInfoBaseTableCell.swift [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    @available(*, unavailable)
+    required init?(coder _: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func setupLayout() {
+        contentView.addSubview(titleLabel)
+        titleLabel.snp.makeConstraints { make in
+            make.leading.top.bottom.equalToSuperview().inset(UIConstants.horizontalInset)
```
Comment (ef1rspb) [inline fearless/Modules/Staking/ValidatorInfo/View/ValidatorInfoBaseTableCell.swift:34]: and where is .trailing?

File: fearless/Modules/Staking/ValidatorInfo/View/ValidatorInfoEmptyStakeCell.swift [hunk 8]
```diff
@@ -0,0 +1,50 @@
+import UIKit
+import SoraUI
+
+final class ValidatorInfoEmptyStakeCell: UITableViewCell {
+    private enum Constants {
+        static let verticalInset: CGFloat = 0
```
Comment (ef1rspb) [inline fearless/Modules/Staking/ValidatorInfo/View/ValidatorInfoEmptyStakeCell.swift:6]: zero inset is just no inset😏
image

File: fearless/Modules/Staking/ValidatorInfo/View/ValidatorInfoStakingAmountCell.swift [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            make.trailing.equalToSuperview().inset(UIConstants.horizontalInset)
+            make.bottom.equalToSuperview().inset(Constants.verticalInset)
+            make.leading.greaterThanOrEqualTo(titleLabel.snp.trailing).offset(UIConstants.horizontalInset)
+        }
+
+        contentView.addSubview(amountLabel)
+        amountLabel.snp.makeConstraints { make in
+            make.trailing.equalToSuperview().inset(UIConstants.horizontalInset)
+            make.top.equalToSuperview().inset(Constants.verticalInset)
+            make.leading.greaterThanOrEqualTo(titleLabel.snp.trailing).offset(UIConstants.horizontalInset)
```
Comment (ef1rspb) [inline fearless/Modules/Staking/ValidatorInfo/View/ValidatorInfoStakingAmountCell.swift:40]: .bottom.equalTo(priceLabel.snp.top)?

File: fearless/Modules/Staking/ValidatorInfo/ViewModel/ValidatorInfoViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+enum ValidatorInfoViewModel {
+    static let headerHeight: CGFloat = 52.0
+    static let rowHeight: CGFloat = 48.0
+    static let accountRowHeight: CGFloat = 56.0
+    static let emptyStakeRowHeight: CGFloat = 140.0
+
+    enum StakingRow {
+        case totalStake(LocalizableResource<StakingAmountViewModel>)
```
Comment (ef1rspb) [inline fearless/Modules/Staking/ValidatorInfo/ViewModel/ValidatorInfoViewModel.swift:39]: LocalizableResource is everywhere. Maybe just pass [LocalizableResource<ValidatorInfoViewModel>] ?

File: fearless/Modules/Staking/ValidatorInfo/ViewModel/ValidatorInfoViewModelFactory.swift [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+        return .account(viewModel)
+    }
+
+    private func createStakingViewModel(
+        from validatorInfo: ValidatorInfoProtocol,
+        priceData: PriceData?
+    ) -> ValidatorInfoViewModel {
+        guard let stakeInfo = validatorInfo.stakeInfo else { return createEmptyStakeRow() }
+
+        var stakingRows: [ValidatorInfoViewModel.StakingRow] = []
```
Comment (ef1rspb) [inline fearless/Modules/Staking/ValidatorInfo/ViewModel/ValidatorInfoViewModelFactory.swift:222]: prefer let over var
```swift
let stakingRows: [ValidatorInfoViewModel.StakingRow] = [
  .nominators(..),
  .totalStake
]

Comment (stas-litvinov) [reply fearless/Modules/Staking/ValidatorInfo/ViewModel/ValidatorInfoViewModelFactory.swift:222]: Sure

File: fearless/Modules/Staking/ValidatorInfo/ViewModel/ValidatorInfoViewModel.swift (hunk 10 above)
Comment (stas-litvinov) [reply fearless/Modules/Staking/ValidatorInfo/ViewModel/ValidatorInfoViewModel.swift:39]: Can't quite understand what you mean. Do you suggest to change view model?

File: fearless/Modules/Staking/ValidatorInfo/View/ValidatorInfoEmptyStakeCell.swift (hunk 8 above)
Comment (stas-litvinov) [reply fearless/Modules/Staking/ValidatorInfo/View/ValidatorInfoEmptyStakeCell.swift:6]: It kinda works 
image

File: fearless/Modules/Staking/ValidatorInfo/View/ValidatorInfoBaseTableCell.swift (hunk 7 above)
Comment (stas-litvinov) [reply fearless/Modules/Staking/ValidatorInfo/View/ValidatorInfoBaseTableCell.swift:34]: Trailing is defined in children ¯\_(ツ)_/¯

File: fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift (hunk 5 above)
Comment (stas-litvinov) [reply fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift:43]: Moved to constants inside VC

File: fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift (hunk 4 above)
Comment (stas-litvinov) [reply fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift:5]: So be it!

File: fearless/Modules/Staking/ValidatorInfo/ValidatorInfoInteractor.swift (hunk 3 above)
Comment (stas-litvinov) [reply fearless/Modules/Staking/ValidatorInfo/ValidatorInfoInteractor.swift:8]: yes

File: Podfile (hunk 1 above)
Comment (stas-litvinov) [reply Podfile:19]: I usually do.. but sometimes I don't ):

File: fearless/Modules/Staking/ValidatorInfo/ViewModel/ValidatorInfoViewModel.swift (hunk 10 above)
Comment (ef1rspb) [reply fearless/Modules/Staking/ValidatorInfo/ViewModel/ValidatorInfoViewModel.swift:39]: `enum ValidatorInfoViewModel` rename to `enum ValidatorInfoSection`
and somewhere in VC:
`var sections: [LocalizableResource<ValidatorInfoSection>] = []`

File: .gitignore [hunk 12]
```diff
@@ -50,6 +50,7 @@ fastlane/
 iOSInjectionProject/
 #
 # Custom scripts
+fearless/env-vars.sh
```
Comment (ERussel) [inline .gitignore:53]: why do we need this?

Comment (stas-litvinov) [reply .gitignore:53]: It contains Moonpay keys. I'll explain it later when Moonpay integration is ready

File: fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift (hunk 6 above)
Comment (ERussel) [reply fearless/Modules/Staking/ValidatorInfo/ValidatorInfoViewController.swift:58]: Nope, but don't see any problems with fixed height cells: more code to write but it already has been written :)

**Review by ERussel — APPROVED** [review]

**Pull request merged**